tutorials/shaders/shader_reference/shader_preprocessor.rst

```diff
@@ -0,0 +1,141 @@
+
+#define
+^^^^^^^
+\ **Syntax:** `#define identifier <replacement_code>`.
+
+Defines the identifier after that directive as macro, and replaces all successive occurrence of it to the replacement code in the shader.
+If the replacement code is not defined, it may only been used within `#ifdef` or `#ifndef` directives.
+
+.. code-block:: glsl
+
+    shader_type spatial;
+
+    #define USE_MY_COLOR
+    #define MY_COLOR vec3(1, 0, 0)
+
+    void fragment() {
+    #ifdef USE_MY_COLOR
+        ALBEDO = MY_COLOR;
+    #endif
+    }
+
+#if
+^^^
+\ **Syntax:** `#if condition`.
+
+The `#if` directive checks the condition and if it evaluates to a non-zero value, the controlled code block is included and skipped otherwise.
+In order to evaluate, the condition must be a simple floating-point or integer expression. There may be multiple condition blocks connected by `||` or `&&` operators.
+It may be continued by `#else` block, but must be ended with `#endif` directive. 
+
+.. code-block:: glsl
+
+    #define VAR 3
+    #define USE_LIGHT 0 // evaluates to false
+    #define USE_COLOR 1 // evaluates to true
+
+    #if VAR == 3 && (USE_LIGHT || USE_COLOR)
+
+
+    #endif
+
+#ifdef
+^^^^^^
+\ **Syntax:** `#ifdef identifier`.
+
+Checks whether the passed identifier is defined by `#define` before that directive. Useful for creating multiple shader versions in the same file.
+It may be continued by `#else` block, but must be ended with `#endif` directive. 
+
+.. code-block:: glsl
+
+    #define USE_LIGHT
+    #ifdef USE_LIGHT
+
+    #endif
+
+#ifndef
+^^^^^^^
+\ **Syntax:** `#ifndef identifier`.
+
+Similar to `#ifdef` but checks whether the passed identifier is not defined by `#define` before that directive.
+
+#else
+^^^^^
+\ **Syntax:** `#else`.
+
+Defines the optional block which is included when the previously defined `#if`, `#ifdef` or `#ifndef` directive is evaluates to false.
+
+.. code-block:: glsl
+
+    shader_type spatial;
+
+    #define MY_COLOR vec3(1, 0, 0)
+
+    void fragment() {
+    #ifndef MY_COLOR
+        ALBEDO = MY_COLOR;
+    #else
+        ALBEDO = vec3(0, 0, 1);
+    #endif
+    }
+
+#endif
+^^^^^^
+\ **Syntax:** `#endif`.
+
+Used as terminator for the `#if`, `#ifdef`, `#ifndef` or subsequent `#else` directives.
+
+#undef
+^^^^^^
+\ **Syntax:** `#undef identifier`.
+
+`#undef directive_name`
+
+The `#undef` directive may be used to cancel the previously defined `#define` directive: 
+
+.. code-block:: glsl
+
+    #define MY_COLOR vec3(1, 0, 0)
+
+    vec3 get_red_color() {
+        return MY_COLOR;
+    }
+
+    #undef MY_COLOR
+    #define MY_COLOR vec3(0, 1, 0)
+
+    vec3 get_green_color() {
+        return MY_COLOR;
+    }
+
+Without `#undef` in that case there will be a macro redefinition error.
+
+#include
+^^^^^^^^
+\ **Syntax:** `#include "path"`.
+
+The `#include` directive includes the content of shader include to a shader. It may be used in any place, but is recommended at the beginning of the shader file, 
```

Comment:
We should make clear what a "shader include" is (a file). Maybe format "shader include" so its clear that this is a thing, and to make the sentence easier to parse.

Reply:
Yes, we can link to the ShaderInclude class reference once the online docs are updated to include it!